### PR TITLE
Added Windows support

### PR DESCRIPTION
**Pull request by kreeuwijk**

### CODE REVIEW

**Comment by kreeuwijk** [comment]
I have tested this with Bolt against Windows 2012 R2:

bolt task run tools::install_windows_agent master=master.inf.puppet.vm --modulepath /root/code --nodes 172.16.219.20 --user vagrant --password vagrant --transport winrm -k

Started on 172.16.219.20...
Finished on 172.16.219.20:
  Saving the install log to C:\Users\Administrator\AppData\Local\Temp\2018-01-11T20-17-27-puppet-install.log
  Installing the Puppet Agent on SERVER2012R2A...
  Notice: /Service[puppet]/ensure: ensure changed 'stopped' to 'running'
  service { 'puppet':
    ensure => 'running',
    enable => 'true',
  }
  Installation has completed.
  Installed
  {
  }
Ran on 1 node in 38.11 seconds

**Comment by MikaelSmith** [comment]
Mind adding a README entry around in https://github.com/puppetlabs/puppetlabs-bootstrap#usage about using it?

@tphoney do you have any suggestions for writing tests for this?

**Comment by kreeuwijk** [comment]
- Added README entry for bootstrap::windows.
- Changed README entry for bootstrap to bootstrap::linux
- Changed init task to linux task
- Changed spec test from init to linux (both name and content)

File: README.md [hunk 1]
```diff
@@ -24,34 +24,39 @@ To run the bootstrap task, use the bolt command, specifying the PE master from w
 
 #### Example: Basic usage
 
-On the command line, run `bolt task run bootstrap master=<master's fqdn> --nodes x,y,z --modules /path/to/modules`
+On the command line:
+* For Linux,   run `bolt task run bootstrap::linux   master=<master's fqdn> --nodes x,y,z --modules /path/to/modules`
+* For Windows, run `bolt task run bootstrap::windows master=<master's fqdn> --nodes x,y,z --modules /path/to/modules`
+For all advanced examples below, simply replace `bootstrap::linux` by `bootstrap::windows` to perform the action on Windows.
```
Comment (MikaelSmith) [inline README.md]: Needs a preceding newline to display properly.

File: README.md [hunk 2]
```diff
@@ -24,34 +24,39 @@ To run the bootstrap task, use the bolt command, specifying the PE master from w
...
 
 #### Example: Verify the master's CA on initial connection
 
-Optionally to validate the connection during the boostrap process, specify the puppet master's CA cert: `bolt task run bootstrap master=<master's fqdn> cacert_content="$(cat /etc/puppetlabs/puppet/ssl/certs/ca.pem)" --nodes x,y,z --modules /path/to/modules`
+Optionally to validate the connection during the bootstrap process, specify the puppet master's CA cert by adding the cacert_content option:
+`bolt task run bootstrap::linux master=<master's fqdn> cacert_content="$(cat /etc/puppetlabs/puppet/ssl/certs/ca.pem)" --nodes x,y,z --modules /path/to/modules`
```
Comment (MikaelSmith) [inline README.md]: If you really want these on a separate line, need to add an extra newline. They'd probably format better with triple ticks, as in
```
bolt task ...
```

File: tasks/windows.ps1 [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+}
+else {
+  $alt_names_arg = ""
+}
+if ($cacert_content) {
+  New-Item -ItemType Directory -Force -Path "$env:ProgramData\PuppetLabs\puppet\etc\ssl\certs"
+  Out-File -InputObject $cacert_content -FilePath "$env:ProgramData\PuppetLabs\puppet\etc\ssl\certs\ca.pem" -Encoding ascii -Force
+}
+
+[Net.ServicePointManager]::ServerCertificateValidationCallback = {$true}
```
Comment (MikaelSmith) [inline tasks/windows.ps1]: This is unfortunate, but I don't see a simple way to make use of the CA cert to validate the download.

Comment (MikaelSmith) [reply tasks/windows.ps1]: https://github.com/PowerShell/PowerShell/issues/4899 might have some ideas.

**Comment by MikaelSmith** [comment]
I'd like to see formatting in the README fixed up. I'd also like @tphoney or someone maintaining this module to weigh in.

File: tasks/windows.ps1 [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+)
+
+Function Validate-Parameter {
+  Param ($param)
+  if ($param -contains "'") {
+    Throw 'Single-quote is not allowed in arguments'
+  }
+}
+
+Validate-Parameter($certname)
```
Comment (Iristyle) [inline tasks/windows.ps1]: This should be a validation attribute for the parameters, not separate Function.

File: tasks/windows.ps1 [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+Validate-Parameter($alt_names)
+
+if ($certname.Length -eq 0) {
+  $certname_arg = "agent:certname='$certname' "
+}
+else {
+  $fqdn = [System.Net.Dns]::GetHostByName(($env:computerName)).Hostname
```
Comment (Iristyle) [inline tasks/windows.ps1]: This is a reasonable start, but this approach can be problematic, especially in environments like AWS and Azure.

File: tasks/windows.ps1 [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+Function Validate-Parameter {
+  Param ($param)
+  if ($param -contains "'") {
+    Throw 'Single-quote is not allowed in arguments'
+  }
+}
+
+Validate-Parameter($certname)
+Validate-Parameter($alt_names)
```
Comment (Iristyle) [inline tasks/windows.ps1]: There is no parameter named `$alt_names`

File: tasks/windows.ps1 [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  $fqdn = [System.Net.Dns]::GetHostByName(($env:computerName)).Hostname 
+  $certname_arg = "agent:certname='$fqdn' "
+}
+if ($dns_alt_names.Length -eq 0) {
+  $alt_names_arg = "agent:dns_alt_names='$alt_names' "
+}
+else {
+  $alt_names_arg = ""
+}
+if ($cacert_content) {
```
Comment (Iristyle) [inline tasks/windows.ps1]: This is going to be a bit tricky... creating the files / folders here has the potential to lay down permissions in a way that either creates a security issue by being too permissive OR prevents Puppet from accessing the files. This should probably be written *after* the agent is installed and those directories have been created... but not sure if that puts us into a chicken and egg situation.

File: tasks/windows.ps1 (hunk 3 above)
Comment (Iristyle) [reply tasks/windows.ps1]: Yes, there are a few things we can do to improve this. I generally prefer not to turn off SSL verification.

File: tasks/windows.ps1 (hunk 4 above)
Comment (glennsarti) [reply tasks/windows.ps1]: Ref ref:
https://blogs.technet.microsoft.com/heyscriptingguy/2011/05/15/simplify-your-powershell-script-with-parameter-validation/

File: tasks/windows.ps1 (hunk 3 above)
Comment (glennsarti) [reply tasks/windows.ps1]: Didn't Jeremy and the gang solve this in the simplified installer?

File: tasks/windows.ps1 [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  New-Item -ItemType Directory -Force -Path "$env:ProgramData\PuppetLabs\puppet\etc\ssl\certs"
+  Out-File -InputObject $cacert_content -FilePath "$env:ProgramData\PuppetLabs\puppet\etc\ssl\certs\ca.pem" -Encoding ascii -Force
+}
+
+[Net.ServicePointManager]::ServerCertificateValidationCallback = {$true}
+$webClient = New-Object System.Net.WebClient
+$webClient.DownloadFile("https://${master}:8140/packages/current/install.ps1", $env:temp + '\install.ps1')
+&($env:temp + '\install.ps1') $certname_arg $alt_names_arg
+Write-Output 'Installed'
```
Comment (glennsarti) [inline tasks/windows.ps1]: Shouldn't there be some validation here that the install.ps1 _actually_ was successful?

File: tasks/windows.ps1 (hunk 7 above)
Comment (glennsarti) [reply tasks/windows.ps1]: Same with the puppet.conf file

**Comment by glennsarti** [comment]
Question: Instead of downloading the install.ps1 from the master why not just revendor it in this module verbatim?  Removes network connectivity issues, download issues and uses all prior art.  Also this task does not allow the user to specify service usernames and passwords for the puppet agent service which is common in Windows organisations.

File: README.md (hunk 1 above)
Comment (glennsarti) [reply README.md]: There are no examples for running this on a Windows host i.e. `--modules /path/to/modules` is not an absolute path on Windows.

File: README.md (hunk 2 above)
Comment (glennsarti) [reply README.md]: Missing example for using on Windows i.e. `cacert_content="$(cat /etc/puppetlabs/puppet/ssl/certs/ca.pem)"` makes no sense on a Windows platform

File: README.md (hunk 1 above)
Comment (glennsarti) [reply README.md]: `For Windows ...` is ambiguous.  That's to install a agent on Windows node, not using this task on a Windows host.

**Comment by glennsarti** [comment]
Should really be updating the metadata here as well too.  Only supporting 2012R2 is really limiting.  2016 is a common deployment option now.

**Comment by MikaelSmith** [comment]
The `install.ps1` script is customized for each PE install. So it's not as simple as copying over.

**Comment by Iristyle** [comment]
Yeah thanks @MikaelSmith - I was going to say something similar. I don't think we can vendor here (and would prefer not to), but might be worth taking a quick look anyhow just to be certain.

**Comment by nmaludy** [comment]
@kreeuwijk @MikaelSmith I'm running Open Source puppet and my puppet server doesn't seem to be hosting a `install.ps1` file. Any thoughts on how to make this work for Open Source?

**Comment by MikaelSmith** [comment]
I don't think any of bootstrap works with Open Source puppet yet. It needs some work to write new install scripts. There is some prior art at https://github.com/puppetlabs/task-modules/tree/master/site/install_puppet/tasks for Linux, but nothing simple to pull over for Windows. The puppet_agent module has some prior art that could be used to build one.

**Comment by Iristyle** [comment]
This is being continued in #18

**Pull request closed without merging**